_posts/2020-07-20-github-review-and-merge.md

```diff
@@ -0,0 +1,131 @@
+that are merged together. Individual commits cannot be merged (unless they
+are wrapped by a PR, in a 1:1 relation).
+
+The merge and rebase methods are preserving the PR commit history intact.
+The squash method on the other hand is combining all PR commits into a single
+commit (recording all commit messages into a single commit message)
+automatically.
+
+The merge method preserves the feature branch, allowing future usage of the
+changes included in it (e.g. backporting or sharing with multiple branches).
+On the other hand, the rebase method places the changes one by one on the
+target branch, keeping a cleaner linear log history but without the changes
+original grouping.
+
+If you value your change history and commit message, do not use squash.
+
+If you intend to merge changes into multiple branches frequently or your
+individual commits are not self contained (i.e. they cannot stand alone without
+the other changes in the PR when tested), use the merge method.
+
+If you want a linear clean change history and your commits are stand alone,
+use the rebase method.
+
+# Track Requested Changes 
+The GitHub review tool has one major disadvantage, especially when compared to
+Gerrit. It is extremely hard to both track changes that respond to review
+comments and keep a proper change history in the project.
+
+The difficulty to track changes in a PR is mainly due to the fact that PR/s
+may include more than one logical change. While helping with reducing tests
+(run once on the top commit instead on each and every one of the commits), it
+does not provide proper means for a reviewer to track his requested changes.
+
+To both support the review process (tracking the changes) and keeping the
+change history in good shape, we can consider the following points:
+
+- Answer each review comment with "Done" or other text, helping the reviewer
+  see that the comment has been addressed.
+- One PR, One commit: Whenever possible, keep one logical commit per PR.
+  - One could use both transient commits to answer review requests or use
+    the pushed changes diffs on the PR discussion page.
+  - When using commits to answer review requests, the commits need to be
+    squashed by the author before a final approval for merge is given.
+- One PR, multiple commits:
+  - If the PR is simple and small enough, using the provided diffs between
+    pushes may be enough for a proper review.
+  - When the PR is not small or when there are multiple requested changes,
+    temporary commits may be used, ordered under the logical commit they
+    belong to. It is up to the author to squash them at the end before
+    the final approval. 
+- When using forced pushes to answer review requests, consider leaving
+  a comment about the changes included int the push.
+- The reviewer can confirm no changes have occurred from his last review by
+  checking the last push diff. This is useful post a squash operation.
+
```

Comment:
Any closing words? You were just stating the reality and whining a little bit. It misses the catharsis.

Reply:
I do not have a lot to add, but it is a good point. I will add a reminder to the intent.